File: app/src/main/java/io/github/finall1008/xiaoaimcp/ui/MainActivity.java
package io.github.finall1008.xiaoaimcp.ui;

import android.app.Activity;
import android.app.AlertDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.Build;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.CompoundButton;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.OptionalInt;

import io.github.finall1008.xiaoaimcp.BridgeApplication;
import io.github.finall1008.xiaoaimcp.BridgeContract;
import io.github.finall1008.xiaoaimcp.R;
import io.github.finall1008.xiaoaimcp.TargetVersionPolicy;
import io.github.finall1008.xiaoaimcp.config.McpServer;
import io.github.finall1008.xiaoaimcp.config.RemoteConfigRepository;
import io.github.libxposed.service.XposedService;

public final class MainActivity extends Activity
        implements BridgeApplication.ServiceStateListener {
    private TextView frameworkStatus;
    private TextView targetStatus;
    private TextView emptyView;
    private Button addButton;
    private ServerAdapter adapter;
    private RemoteConfigRepository repository;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setTitle("超级小爱 MCP Bridge");
        buildUi();
    }

    @Override
    protected void onStart() {
        super.onStart();
        BridgeApplication.addServiceStateListener(this, true);
    }

    @Override
    protected void onResume() {
        super.onResume();
        refresh();
    }

    @Override
    protected void onStop() {
        BridgeApplication.removeServiceStateListener(this);
        super.onStop();
    }

    @Override
    public void onServiceStateChanged(XposedService service) {
        runOnUiThread(this::refresh);
    }

    private void buildUi() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setPadding(UiSupport.dp(this, 20), UiSupport.dp(this, 20),
                UiSupport.dp(this, 20), UiSupport.dp(this, 16));
        root.setBackgroundColor(Color.WHITE);

        root.addView(UiSupport.title(this, "超级小爱 MCP Bridge"), UiSupport.matchWrap());
        TextView subtitle = UiSupport.text(this,
                "为超级小爱 8.0 及以上版本注入 Streamable HTTP / SSE MCP 服务器。",
                14);
        subtitle.setTextColor(Color.rgb(90, 95, 108));
        subtitle.setPadding(0, 0, 0, UiSupport.dp(this, 14));
        root.addView(subtitle, UiSupport.matchWrap());

        frameworkStatus = statusView();
        targetStatus = statusView();
        root.addView(frameworkStatus, UiSupport.matchWrap());
        root.addView(targetStatus, UiSupport.matchWrap());

        addButton = new Button(this);
        addButton.setText(R.string.add_server);
        addButton.setAllCaps(false);
        addButton.setOnClickListener(v -> startActivity(
                new Intent(this, ServerEditActivity.class)));
        LinearLayout.LayoutParams buttonParams = UiSupport.matchWrap();
        buttonParams.topMargin = UiSupport.dp(this, 12);
        root.addView(addButton, buttonParams);

        emptyView = UiSupport.text(this, "尚未添加第三方 MCP 服务器", 15);
        emptyView.setGravity(Gravity.CENTER);
        emptyView.setTextColor(Color.rgb(120, 125, 138));
        emptyView.setPadding(0, UiSupport.dp(this, 40), 0, UiSupport.dp(this, 40));

        ListView listView = new ListView(this);
        adapter = new ServerAdapter();
        listView.setAdapter(adapter);
        listView.setDividerHeight(0);
        listView.setEmptyView(emptyView);
        root.addView(emptyView, UiSupport.matchWrap());
        LinearLayout.LayoutParams listParams = new LinearLayout.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                0,
                1
        );
        root.addView(listView, listParams);
        setContentView(root);
    }

    private TextView statusView() {
        TextView view = UiSupport.text(this, "", 13);
        view.setPadding(UiSupport.dp(this, 10), UiSupport.dp(this, 7),
                UiSupport.dp(this, 10), UiSupport.dp(this, 7));
        return view;
    }

    private void refresh() {
        XposedService service = BridgeApplication.service();
        boolean serviceReady = false;
        if (service == null) {
            setStatus(frameworkStatus, false, "API 102 服务未连接：请安装、启用模块并重新打开此页面");
        } else {
            try {
                int api = service.getApiVersion();
                serviceReady = api >= XposedService.API_102;
                setStatus(frameworkStatus, serviceReady,
                        service.getFrameworkName() + " " + service.getFrameworkVersion()
                                + " · API " + api);
            } catch (RuntimeException error) {
                setStatus(frameworkStatus, false, "读取 Xposed 服务失败：" + safeMessage(error));
            }
        }

        boolean targetReady = updateTargetStatus();
        boolean editingEnabled = serviceReady && targetReady;
        addButton.setEnabled(editingEnabled);
        adapter.setEditingEnabled(editingEnabled);

        SharedPreferences preferences = serviceReady ? BridgeApplication.remotePreferences() : null;
        repository = preferences == null ? null : new RemoteConfigRepository(preferences);
        if (repository == null) {
            adapter.replace(List.of());
            return;
        }
        try {
            adapter.replace(repository.load());
        } catch (Exception error) {
            adapter.replace(List.of());
            setStatus(frameworkStatus, false, "MCP 配置损坏：" + safeMessage(error));
        }
    }

    private boolean updateTargetStatus() {
        try {
            @SuppressWarnings("deprecation")
            PackageInfo info = getPackageManager().getPackageInfo(
                    BridgeContract.TARGET_PACKAGE,
                    0
            );
            @SuppressWarnings("deprecation")
            long code = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P
                    ? info.getLongVersionCode()
                    : (info.versionCode & 0xffffffffL);
            String versionName = info.versionName;
            boolean supported = TargetVersionPolicy.isSupported(versionName);
            boolean reference = supported
                    && BridgeContract.REFERENCE_VERSION_NAME.equals(versionName)
                    && code == BridgeContract.REFERENCE_VERSION_CODE;
            String version = "超级小爱 " + String.valueOf(versionName) + " (" + code + ")";
            if (reference) {
                setStatus(targetStatus, StatusLevel.SUCCESS, version + " · 已验证");
            } else if (supported) {
                setStatus(targetStatus, StatusLevel.WARNING,
                        version + " · 将在目标进程中自动探测兼容性");
            } else {
                OptionalInt major = TargetVersionPolicy.parseMajor(versionName);
                String reason = major.isPresent()
                        ? "要求超级小爱 8.0 或更高版本"
                        : "无法识别版本号，要求超级小爱 8.0 或更高版本";
                setStatus(targetStatus, StatusLevel.ERROR, version + " · " + reason);
            }
            return supported;
        } catch (PackageManager.NameNotFoundException error) {
            setStatus(targetStatus, false, "未安装超级小爱：" + BridgeContract.TARGET_PACKAGE);
            return false;
        }
    }

    private void setStatus(TextView view, boolean success, String text) {
        setStatus(view, success ? StatusLevel.SUCCESS : StatusLevel.ERROR, text);
    }

    private void setStatus(TextView view, StatusLevel level, String text) {
        if (level == StatusLevel.SUCCESS) {
            view.setText(getString(R.string.status_success, text));
            view.setTextColor(Color.rgb(27, 125, 75));
            view.setBackgroundColor(Color.rgb(235, 248, 240));
        } else if (level == StatusLevel.WARNING) {
            view.setText(getString(R.string.status_warning, text));
            view.setTextColor(Color.rgb(145, 96, 16));
            view.setBackgroundColor(Color.rgb(255, 247, 226));
        } else {
            view.setText(getString(R.string.status_error, text));
            view.setTextColor(Color.rgb(180, 62, 52));
            view.setBackgroundColor(Color.rgb(253, 239, 237));
        }
    }

    private void edit(McpServer server) {
        Intent intent = new Intent(this, ServerEditActivity.class);
        intent.putExtra(ServerEditActivity.EXTRA_SERVER_ID, server.id());
        startActivity(intent);
    }

    private void delete(McpServer server) {
        new AlertDialog.Builder(this)
                .setTitle("删除服务器")
                .setMessage("确认删除 “" + server.name() + "”？超级小爱中的对应工具会在线注销。")
                .setNegativeButton("取消", null)
                .setPositiveButton("删除", (dialog, which) -> {
                    try {
                        requireRepository().delete(server.id());
                        refresh();
                    } catch (Exception error) {
                        showError(error);
                    }
                })
                .show();
    }

    private RemoteConfigRepository requireRepository() {
        if (repository == null) {
            throw new IllegalStateException("API 102 服务未连接");
        }
        return repository;
    }

    private void showError(Throwable error) {
        Toast.makeText(this, safeMessage(error), Toast.LENGTH_LONG).show();
    }

    private static String safeMessage(Throwable error) {
        String message = error.getMessage();
        return message == null || message.isBlank() ? error.getClass().getSimpleName() : message;
    }

    private final class ServerAdapter extends BaseAdapter {
        private final List<McpServer> servers = new ArrayList<>();
        private boolean editingEnabled;

        void setEditingEnabled(boolean enabled) {
            if (editingEnabled != enabled) {
                editingEnabled = enabled;
                notifyDataSetChanged();
            }
        }

        void replace(List<McpServer> updated) {
            servers.clear();
            servers.addAll(updated);
            notifyDataSetChanged();
        }

        @Override
        public int getCount() {
            return servers.size();
        }

        @Override
        public McpServer getItem(int position) {
            return servers.get(position);
        }

        @Override
        public long getItemId(int position) {
            return getItem(position).id().hashCode();
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            McpServer server = getItem(position);
            LinearLayout row = new LinearLayout(MainActivity.this);
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setGravity(Gravity.CENTER_VERTICAL);
            row.setPadding(UiSupport.dp(MainActivity.this, 12), UiSupport.dp(MainActivity.this, 12),
                    UiSupport.dp(MainActivity.this, 8), UiSupport.dp(MainActivity.this, 12));
            row.setBackgroundColor(Color.rgb(247, 248, 251));

            LinearLayout copy = new LinearLayout(MainActivity.this);
            copy.setOrientation(LinearLayout.VERTICAL);
            TextView title = UiSupport.text(MainActivity.this, server.name(), 17);
            title.setTypeface(Typeface.DEFAULT, Typeface.BOLD);
            copy.addView(title, UiSupport.matchWrap());
            String detail = server.transport().toUpperCase(Locale.ROOT) + " · " + server.url()
                    + (server.headers().isEmpty() ? "" : " · " + server.headers().size() + " 个请求头");
            TextView details = UiSupport.text(MainActivity.this, detail, 12);
            details.setTextColor(Color.rgb(100, 105, 118));
            details.setMaxLines(2);
            copy.addView(details, UiSupport.matchWrap());
            LinearLayout.LayoutParams copyParams = new LinearLayout.LayoutParams(0,
                    ViewGroup.LayoutParams.WRAP_CONTENT, 1);
            row.addView(copy, copyParams);

            Switch enabled = new Switch(MainActivity.this);
            enabled.setChecked(server.enabled());
            enabled.setContentDescription("启用 " + server.name());
            enabled.setEnabled(editingEnabled);
            if (editingEnabled) {
                enabled.setOnCheckedChangeListener((CompoundButton button, boolean checked) -> {
                    if (checked == server.enabled()) {
                        return;
                    }
                    try {
                        requireRepository().setEnabled(server.id(), checked);
                        refresh();
                    } catch (Exception error) {
                        button.setOnCheckedChangeListener(null);
                        button.setChecked(server.enabled());
                        showError(error);
                    }
                });
            }
            row.addView(enabled, new LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
            ));
            if (editingEnabled) {
                row.setOnClickListener(v -> edit(server));
                row.setOnLongClickListener(v -> {
                    delete(server);
                    return true;
                });
            }

            LinearLayout wrapper = new LinearLayout(MainActivity.this);
            wrapper.setPadding(0, UiSupport.dp(MainActivity.this, 5), 0,
                    UiSupport.dp(MainActivity.this, 5));
            wrapper.addView(row, UiSupport.matchWrap());
            return wrapper;
        }
    }

    private enum StatusLevel {
        SUCCESS,
        WARNING,
        ERROR
    }
}

File: app/src/main/java/io/github/finall1008/xiaoaimcp/ui/MainActivity.kt
package io.github.finall1008.xiaoaimcp.ui

import android.content.ClipData
import android.content.ClipboardManager
import android.content.SharedPreferences
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.unit.dp
import io.github.finall1008.xiaoaimcp.BridgeApplication
import io.github.finall1008.xiaoaimcp.BridgeContract
import io.github.finall1008.xiaoaimcp.BuildConfig
import io.github.finall1008.xiaoaimcp.R
import io.github.finall1008.xiaoaimcp.TargetVersionPolicy
import io.github.finall1008.xiaoaimcp.config.LegacyMcpConfigRepository
import io.github.finall1008.xiaoaimcp.filepolicy.FilePolicyConfig
import io.github.finall1008.xiaoaimcp.filepolicy.FilePolicyRepository
import io.github.finall1008.xiaoaimcp.restart.RestartStateListener
import io.github.finall1008.xiaoaimcp.restart.XiaoAiRootRestarter
import io.github.finall1008.xiaoaimcp.timeout.FirstOutputTimeoutConfig
import io.github.finall1008.xiaoaimcp.timeout.FirstOutputTimeoutMode
import io.github.finall1008.xiaoaimcp.timeout.FirstOutputTimeoutRepository
import io.github.libxposed.service.XposedService
import top.yukonga.miuix.kmp.basic.BasicComponent
import top.yukonga.miuix.kmp.basic.Card
import top.yukonga.miuix.kmp.basic.FloatingNavigationBar
import top.yukonga.miuix.kmp.basic.FloatingNavigationBarItem
import top.yukonga.miuix.kmp.basic.Icon
import top.yukonga.miuix.kmp.basic.IconButton
import top.yukonga.miuix.kmp.basic.ScrollBehavior
import top.yukonga.miuix.kmp.basic.Text
import top.yukonga.miuix.kmp.icon.MiuixIcons
import top.yukonga.miuix.kmp.icon.extended.Home
import top.yukonga.miuix.kmp.icon.extended.Info
import top.yukonga.miuix.kmp.icon.extended.Link
import top.yukonga.miuix.kmp.icon.extended.ListView
import top.yukonga.miuix.kmp.icon.extended.Lock
import top.yukonga.miuix.kmp.icon.extended.Messages
import top.yukonga.miuix.kmp.icon.extended.Refresh
import top.yukonga.miuix.kmp.icon.extended.Timer
import top.yukonga.miuix.kmp.preference.ArrowPreference
import top.yukonga.miuix.kmp.theme.MiuixTheme

class MainActivity :
    ComponentActivity(),
    BridgeApplication.ServiceStateListener,
    RestartStateListener {
    private var state by mutableStateOf(MainUiState())
    private var errorMessage by mutableStateOf<String?>(null)
    private var preferences: SharedPreferences? = null
    private var legacyMcpRepository: LegacyMcpConfigRepository? = null
    private var restartInFlight by mutableStateOf(false)
    private var legacyMigrationNoticePending by mutableStateOf(false)
    private val agentTraceController by lazy { AgentTraceUiController(this) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContent {
            BridgeTheme(this) {
                MainScreen(
                    state = state,
                    errorMessage = errorMessage,
                    onDismissError = {
                        XiaoAiRootRestarter.clearError()
                        errorMessage = null
                    },
                    onOpenFilePolicy = {
                        startActivity(Intent(this, FilePolicyActivity::class.java))
                    },
                    onOpenPromptPatches = {
                        startActivity(Intent(this, PromptPatchActivity::class.java))
                    },
                    agentTraceController = agentTraceController,
                    onOpenFirstOutputTimeout = {
                        startActivity(Intent(this, FirstOutputTimeoutActivity::class.java))
                    },
                    restartInFlight = restartInFlight,
                    onRestartXiaoAi = ::restartXiaoAi,
                    onOpenReleases = ::openReleases,
                    showLegacyMcpMigrationNotice = legacyMigrationNoticePending,
                    onDismissLegacyMcpMigrationNotice = ::dismissLegacyMcpMigrationNotice,
                    onCopyLegacyMcpConfig = ::copyLegacyMcpConfig,
                )
            }
        }
    }

    override fun onStart() {
        super.onStart()
        BridgeApplication.addServiceStateListener(this, true)
        XiaoAiRootRestarter.addListener(this)
        agentTraceController.start()
    }

    override fun onResume() {
        super.onResume()
        refresh()
        agentTraceController.resume()
    }

    override fun onStop() {
        XiaoAiRootRestarter.removeListener(this)
        BridgeApplication.removeServiceStateListener(this)
        agentTraceController.stop()
        super.onStop()
    }

    override fun onDestroy() {
        agentTraceController.destroy()
        super.onDestroy()
    }

    override fun onServiceStateChanged(service: XposedService?) {
        runOnUiThread(::refresh)
    }

    override fun onRestartStateChanged(inFlight: Boolean, errorMessage: String?) {
        runOnUiThread {
            restartInFlight = inFlight
            if (errorMessage != null) this.errorMessage = errorMessage
        }
    }

    private fun refresh() {
        val service = BridgeApplication.service()
        var serviceReady = false
        var frameworkStatus = if (service == null) {
            StatusUi(StatusLevel.ERROR, "API 102 服务未连接：请安装、启用模块并重新打开此页面")
        } else {
            try {
                val api = service.apiVersion
                serviceReady = api >= XposedService.API_102
                StatusUi(
                    if (serviceReady) StatusLevel.SUCCESS else StatusLevel.ERROR,
                    "${service.frameworkName} ${service.frameworkVersion} · API $api",
                )
            } catch (error: RuntimeException) {
                StatusUi(StatusLevel.ERROR, "读取 Xposed 服务失败：${safeMessage(error)}")
            }
        }

        val (targetStatus, targetReady, nativeMcpAvailable) = targetStatus()
        preferences = if (serviceReady) {
            BridgeApplication.remotePreferences()
        } else {
            null
        }
        legacyMcpRepository = preferences?.let(::LegacyMcpConfigRepository)
        val legacyServers = try {
            legacyMcpRepository?.load().orEmpty()
        } catch (error: Exception) {
            frameworkStatus = StatusUi(
                StatusLevel.ERROR,
                "旧版 MCP 配置损坏，无法导出：${safeMessage(error)}",
            )
            emptyList()
        }
        legacyMigrationNoticePending = legacyServers.isNotEmpty() &&
            nativeMcpAvailable &&
            preferences?.getBoolean(
                BridgeContract.PREF_LEGACY_MCP_MIGRATION_NOTICE_SEEN,
                false,
            ) != true
        val filePolicySummary = if (preferences == null) {
            "API 102 服务未连接"
        } else {
            try {
                filePolicySummary(FilePolicyRepository(requireNotNull(preferences)).load())
            } catch (error: RuntimeException) {
                "配置损坏：${safeMessage(error)}"
            }
        }
        state = MainUiState(
            frameworkStatus = frameworkStatus,
            targetStatus = targetStatus,
            editingEnabled = serviceReady && targetReady,
            firstOutputTimeout = preferences?.let {
                FirstOutputTimeoutRepository(it).load()
            } ?: FirstOutputTimeoutConfig.hostDefault(),
            filePolicySummary = filePolicySummary,
            legacyMcpServerCount = legacyServers.size,
            nativeMcpAvailable = nativeMcpAvailable,
        )
    }

    @Suppress("DEPRECATION")
    private fun targetStatus(): Triple<StatusUi, Boolean, Boolean> {
        return try {
            val info = packageManager.getPackageInfo(BridgeContract.TARGET_PACKAGE, 0)
            val code = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                info.longVersionCode
            } else {
                info.versionCode.toLong() and 0xffffffffL
            }
            val versionName = info.versionName
            val supported = TargetVersionPolicy.isSupported(versionName)
            val reference = supported &&
                BridgeContract.REFERENCE_VERSION_NAME == versionName &&
                BridgeContract.REFERENCE_VERSION_CODE == code
            val version = "超级小爱 ${versionName.toString()} ($code)"
            val nativeMcp = TargetVersionPolicy.hasNativeMcp(versionName)
            when {
                !nativeMcp && supported -> StatusUi(
                    StatusLevel.WARNING,
                    "$version · 增强功能可用；原生 MCP 需 ${BridgeContract.NATIVE_MCP_VERSION_NAME}+",
                ).withReadiness(true, false)
                reference -> StatusUi(StatusLevel.SUCCESS, "$version · 已验证")
                    .withReadiness(true, true)
                supported -> StatusUi(
                    StatusLevel.WARNING,
                    "$version · 将在目标进程中自动探测兼容性",
                ).withReadiness(true, true)
                else -> {
                    val reason = if (TargetVersionPolicy.parseMajor(versionName).isPresent) {
                        "要求超级小爱 8.0 或更高版本"
                    } else {
                        "无法识别版本号，要求超级小爱 8.0 或更高版本"
                    }
                    StatusUi(StatusLevel.ERROR, "$version · $reason")
                        .withReadiness(false, false)
                }
            }
        } catch (_: PackageManager.NameNotFoundException) {
            StatusUi(StatusLevel.ERROR, "未安装超级小爱：${BridgeContract.TARGET_PACKAGE}")
                .withReadiness(false, false)
        }
    }

    private fun copyLegacyMcpConfig() {
        try {
            val json = legacyMcpRepository?.exportNativeConfig()
                ?: error("API 102 服务未连接")
            val clipboard = getSystemService(ClipboardManager::class.java)
                ?: error("系统剪贴板不可用")
            clipboard.setPrimaryClip(ClipData.newPlainText("超级小爱 MCP 配置", json))
            Toast.makeText(
                this,
                "已复制，请粘贴到超级小爱 → 设置 → MCP 服务",
                Toast.LENGTH_LONG,
            ).show()
            dismissLegacyMcpMigrationNotice()
        } catch (error: Exception) {
            errorMessage = safeMessage(error)
        }
    }

    private fun openReleases() {
        try {
            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(GITHUB_RELEASES_URL)))
        } catch (error: Exception) {
            errorMessage = "无法打开 GitHub Releases：${safeMessage(error)}"
        }
    }

    private fun restartXiaoAi() {
        XiaoAiRootRestarter.restart(applicationContext)
    }

    private fun dismissLegacyMcpMigrationNotice() {
        preferences?.edit()
            ?.putBoolean(BridgeContract.PREF_LEGACY_MCP_MIGRATION_NOTICE_SEEN, true)
            ?.apply()
        legacyMigrationNoticePending = false
    }
}

private data class MainUiState(
    val frameworkStatus: StatusUi = StatusUi(StatusLevel.ERROR, "正在连接 API 102 服务…"),
    val targetStatus: StatusUi = StatusUi(StatusLevel.WARNING, "正在检查超级小爱版本…"),
    val editingEnabled: Boolean = false,
    val firstOutputTimeout: FirstOutputTimeoutConfig =
        FirstOutputTimeoutConfig.hostDefault(),
    val filePolicySummary: String = "正在读取文件权限配置…",
    val legacyMcpServerCount: Int = 0,
    val nativeMcpAvailable: Boolean = false,
)

private data class StatusUi(val level: StatusLevel, val text: String)

private enum class StatusLevel { SUCCESS, WARNING, ERROR }

internal enum class RootPage(val label: String) {
    HOME("首页"),
    TRACE("轨迹"),
    ABOUT("关于"),
}

internal val ROOT_PAGES = RootPage.entries.toList()
internal val DEFAULT_ROOT_PAGE = RootPage.HOME
internal const val GITHUB_RELEASES_URL =
    "https://github.com/finall1008/hyper-ai-third-party-mcp/releases"

@Composable
private fun MainScreen(
    state: MainUiState,
    errorMessage: String?,
    onDismissError: () -> Unit,
    onOpenFilePolicy: () -> Unit,
    onOpenPromptPatches: () -> Unit,
    agentTraceController: AgentTraceUiController,
    onOpenFirstOutputTimeout: () -> Unit,
    restartInFlight: Boolean,
    onRestartXiaoAi: () -> Unit,
    onOpenReleases: () -> Unit,
    showLegacyMcpMigrationNotice: Boolean,
    onDismissLegacyMcpMigrationNotice: () -> Unit,
    onCopyLegacyMcpConfig: () -> Unit,
) {
    var showRestartConfirmation by remember { mutableStateOf(false) }
    var selectedPage by rememberSaveable { mutableStateOf(DEFAULT_ROOT_PAGE) }
    val appName = stringResource(R.string.app_name)
    Box(modifier = Modifier.fillMaxSize()) {
        key(selectedPage) {
            when (selectedPage) {
                RootPage.TRACE -> AgentTraceRoute(
                    controller = agentTraceController,
                    onBack = null,
                    extraBottomPadding = 108.dp,
                )
                RootPage.HOME, RootPage.ABOUT -> BridgePageScaffold(
                    title = if (selectedPage == RootPage.HOME) appName else "关于",
                    actions = {
                        if (selectedPage == RootPage.HOME) {
                            IconButton(
                                onClick = { showRestartConfirmation = true },
                                enabled = !restartInFlight,
                            ) {
                                Icon(
                                    imageVector = MiuixIcons.Refresh,
                                    contentDescription = "重启超级小爱",
                                    tint = MiuixTheme.colorScheme.onSurface,
                                )
                            }
                        }
                    },
                ) { padding, scrollBehavior ->
                    if (selectedPage == RootPage.HOME) {
                        HomePage(
                            state = state,
                            scaffoldPadding = padding,
                            scrollBehavior = scrollBehavior,
                            onOpenFilePolicy = onOpenFilePolicy,
                            onOpenPromptPatches = onOpenPromptPatches,
                            onEditTimeout = onOpenFirstOutputTimeout,
                            onCopyLegacyMcpConfig = onCopyLegacyMcpConfig,
                        )
                    } else {
                        AboutPage(
                            scaffoldPadding = padding,
                            scrollBehavior = scrollBehavior,
                            onOpenReleases = onOpenReleases,
                        )
                    }
                }
            }
        }
        Box(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .navigationBarsPadding()
                .padding(bottom = 12.dp),
            contentAlignment = Alignment.Center,
        ) {
            FloatingNavigationBar {
                FloatingNavigationBarItem(
                    selected = selectedPage == RootPage.HOME,
                    onClick = { selectedPage = RootPage.HOME },
                    icon = MiuixIcons.Home,
                    label = RootPage.HOME.label,
                )
                FloatingNavigationBarItem(
                    selected = selectedPage == RootPage.TRACE,
                    onClick = { selectedPage = RootPage.TRACE },
                    icon = MiuixIcons.ListView,
                    label = RootPage.TRACE.label,
                )
                FloatingNavigationBarItem(
                    selected = selectedPage == RootPage.ABOUT,
                    onClick = { selectedPage = RootPage.ABOUT },
                    icon = MiuixIcons.Info,
                    label = RootPage.ABOUT.label,
                )
            }
        }
    }
    ConfirmDialog(
        show = showRestartConfirmation,
        title = "重启超级小爱",
        message = "需要 root 权限，并将结束超级小爱全部进程后重新打开。正在进行的对话、模型响应和工具调用会立即中断。",
        confirmText = "重启",
        onDismiss = { showRestartConfirmation = false },
        onConfirm = {
            showRestartConfirmation = false
            onRestartXiaoAi()
        },
    )
    ConfirmDialog(
        show = showLegacyMcpMigrationNotice,
        title = "迁移旧版 MCP 配置",
        message = "模块已停用第三方 MCP 注入，检测到旧配置。可以复制为超级小爱原生 JSON；内容可能包含请求头和访问凭据，请勿分享。",
        confirmText = "复制配置",
        onDismiss = onDismissLegacyMcpMigrationNotice,
        onConfirm = onCopyLegacyMcpConfig,
    )
    MessageDialog(
        message = errorMessage,
        title = "操作失败",
        onDismiss = onDismissError,
    )
}

@Composable
private fun HomePage(
    state: MainUiState,
    scaffoldPadding: PaddingValues,
    scrollBehavior: ScrollBehavior,
    onOpenFilePolicy: () -> Unit,
    onOpenPromptPatches: () -> Unit,
    onEditTimeout: () -> Unit,
    onCopyLegacyMcpConfig: () -> Unit,
) {
    BridgePageList(
        scaffoldPadding = scaffoldPadding,
        scrollBehavior = scrollBehavior,
        extraBottomPadding = 108.dp,
    ) {
        item { SectionLabel("运行状态") }
        item {
            Card(modifier = Modifier.fillMaxWidth()) {
                StatusRow("Xposed 框架", state.frameworkStatus)
                StatusRow("目标应用", state.targetStatus)
            }
        }
        item { SectionLabel("Agent 设置", Modifier.padding(top = 4.dp)) }
        item {
            Card(modifier = Modifier.fillMaxWidth()) {
                ArrowPreference(
                    title = "模型首次输出超时",
                    summary = firstOutputTimeoutSummary(state.firstOutputTimeout),
                    startAction = { PreferenceIcon(MiuixIcons.Timer) },
                    enabled = state.editingEnabled,
                    onClick = onEditTimeout,
                )
                ArrowPreference(
                    title = "System Prompt 补丁",
                    summary = "精确替换规则与变更预览",
                    startAction = { PreferenceIcon(MiuixIcons.Messages) },
                    enabled = state.editingEnabled,
                    onClick = onOpenPromptPatches,
                )
            }
        }
        item { SectionLabel("权限设置", Modifier.padding(top = 4.dp)) }
        item {
            Card(modifier = Modifier.fillMaxWidth()) {
                ArrowPreference(
                    title = "文件访问权限",
                    summary = state.filePolicySummary,
                    startAction = { PreferenceIcon(MiuixIcons.Lock) },
                    enabled = state.editingEnabled,
                    onClick = onOpenFilePolicy,
                )
            }
        }
        if (state.legacyMcpServerCount > 0) {
            item { SectionLabel("旧版 MCP 配置", Modifier.padding(top = 4.dp)) }
            item {
                Card(modifier = Modifier.fillMaxWidth()) {
                    ArrowPreference(
                        title = "复制原生 MCP JSON",
                        summary = if (state.nativeMcpAvailable) {
                            "${state.legacyMcpServerCount} 个旧服务器；模块不再注入，请迁移到超级小爱设置"
                        } else {
                            "${state.legacyMcpServerCount} 个旧服务器；升级到 ${BridgeContract.NATIVE_MCP_VERSION_NAME}+ 后迁移"
                        },
                        startAction = { PreferenceIcon(MiuixIcons.Link) },
                        enabled = state.editingEnabled && state.nativeMcpAvailable,
                        onClick = onCopyLegacyMcpConfig,
                    )
                    BasicComponent(
                        title = "迁移位置",
                        summary = "超级小爱 → 设置 → MCP 服务；剪贴板内容可能包含访问凭据",
                        startAction = { PreferenceIcon(MiuixIcons.Info) },
                        enabled = false,
                    )
                }
            }
        }
    }
}

@Composable
private fun StatusRow(title: String, status: StatusUi) {
    val (label, color) = when (status.level) {
        StatusLevel.SUCCESS -> "正常" to MiuixTheme.colorScheme.primary
        StatusLevel.WARNING -> "注意" to MiuixTheme.colorScheme.onTertiaryContainer
        StatusLevel.ERROR -> "异常" to MiuixTheme.colorScheme.error
    }
    BasicComponent(
        title = title,
        summary = status.text,
        startAction = { PreferenceIcon(MiuixIcons.Info) },
        endActions = { Text(label, style = MiuixTheme.textStyles.body2, color = color) },
    )
}

@Composable
private fun AboutPage(
    scaffoldPadding: PaddingValues,
    scrollBehavior: ScrollBehavior,
    onOpenReleases: () -> Unit,
) {
    BridgePageList(
        scaffoldPadding = scaffoldPadding,
        scrollBehavior = scrollBehavior,
        extraBottomPadding = 108.dp,
    ) {
        item {
            Card(modifier = Modifier.fillMaxWidth()) {
                BasicComponent(
                    title = stringResource(R.string.app_name),
                    summary = aboutVersionLabel(BuildConfig.VERSION_NAME),
                    startAction = { PreferenceIcon(MiuixIcons.Info) },
                )
                ArrowPreference(
                    title = "GitHub Releases",
                    summary = "查看版本更新与安装包",
                    startAction = { PreferenceIcon(MiuixIcons.Link) },
                    onClick = onOpenReleases,
                )
            }
        }
    }
}

internal fun firstOutputTimeoutSummary(config: FirstOutputTimeoutConfig): String {
    return when (config.mode()) {
        FirstOutputTimeoutMode.HOST_DEFAULT -> "跟随宿主（参考版 120 秒）"
        FirstOutputTimeoutMode.CUSTOM -> "自定义 ${config.customSeconds()} 秒"
        FirstOutputTimeoutMode.UNLIMITED -> "不限制"
    }
}

internal fun filePolicySummary(config: FilePolicyConfig): String {
    val enabled = if (config.enabled()) "已启用" else "未启用"
    val rules = if (config.rules().isEmpty()) {
        "尚未配置目录规则"
    } else {
        "${config.rules().size} 条目录规则"
    }
    return "$enabled · $rules"
}

internal fun aboutVersionLabel(versionName: String): String {
    return "当前版本 $versionName"
}

private fun StatusUi.withReadiness(
    targetReady: Boolean,
    nativeMcpAvailable: Boolean,
): Triple<StatusUi, Boolean, Boolean> {
    return Triple(this, targetReady, nativeMcpAvailable)
}
